src/web/js/output-ui.js

```diff
@@ -612,70 +394,46 @@
-              line: cmloc.end.line,
-              ch: editors[cmloc.source].getLine(cmloc.end.line).length}};
-        }
-        return editors[cmloc.source].getRange(cmloc.start, cmloc.end);
-      } else {
-        var source = sessionStorage.getItem(cmloc.source);
-        if(!source) {
-          return undefined;
-        }
-        var lines = source.split("\n").slice(cmloc.start.line, cmloc.end.line + 1);
-        if(tight) {
-          lines[lines.length - 1] = lines[lines.length - 1].substr(0, cmloc.end.ch);
-          lines[0] = lines[0].substr(cmloc.start.ch);
-        }
-        return lines.join("\n");
-      }
-    }
-    
-    function makeMaybeLocToAST(runtime, editors, srcloc) {
+    function makeMaybeLocToAST(runtime, documents, srcloc) {
       return runtime.makeFunction(function(loc) {
         return runtime.ffi.cases(runtime.getField(srcloc, "is-Srcloc"), "Srcloc", loc, {
           "builtin": function(_) {
             console.error("maybeLocToAST should not be passed a builtin source location.", loc);
             return runtime.ffi.makeNone();
           },
-          "srcloc": function(filename, start_line, start_col, startCh, endL, endC, endCh) {
+          "srcloc": function(filename, start_line, start_col, _, end_line, end_col, _) {
             var prelude = ""
             for(var i=1; i < start_line; i++) {prelude += "\n";}
             for(var i=0; i < start_col; i++)  {prelude += " "; }
-            var source = getSourceContent(editors, cmPosFromSrcloc(runtime, srcloc, loc), true);
-            if(source === undefined) {
+            if(!documents.has(filename))
               return runtime.ffi.makeNone();
-            } else {
-              runtime.pauseStack(function(restarter) {
-                runtime.run(function(_, __) {
-                  return runtime.getField(PP, "surface-parse").app(prelude + source, filename);
-                }, runtime.namespace, {
-                  sync: false
-                }, function(result) {
-                  if(runtime.isSuccessResult(result)) {
-                    restarter.resume(runtime.ffi.makeSome(result.result.dict.block.dict.stmts.dict.first));
-                  } else {
-                    restarter.resume(runtime.ffi.makeNone());
-                  }
-                });
+            var source = documents.
+              get(filename).
+              getRange(
+                new CodeMirror.Pos(start_line - 1, start_col),
+                new CodeMirror.Pos(  end_line - 1,   end_col));
+            runtime.pauseStack(function(restarter) {
+              runtime.run(function(_, __) {
+                return runtime.getField(PP, "surface-parse").app(prelude + source, filename);
+              }, runtime.namespace, {
+                sync: false
+              }, function(result) {
+                if(runtime.isSuccessResult(result)) {
+                  restarter.resume(runtime.ffi.makeSome(result.result.dict.block.dict.stmts.dict.first));
```

Comment:
Also, this fails badly on the empty program.  I know that the empty program "can't possibly have any errors," except I'm sure there's some way to make it happen by messing up default includes, so we should make sure that doesn't fudge up.